core/services/eth/node.go

```diff
@@ -0,0 +1,323 @@
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/pkg/errors"
+	"github.com/smartcontractkit/chainlink/core/logger"
+)
+
+type (
+	// node represents one ethereum node.
+	// It must have a ws url and may have a http url
+	node struct {
+		ws      rawclient
+		http    *rawclient
+		log     *logger.Logger
+		dialled bool
+	}
+
+	rawclient struct {
+		rpc  *rpc.Client
+		geth *ethclient.Client
+		uri  url.URL
+	}
+)
+
+func newNode(wsuri url.URL, httpuri *url.URL, name string) (n *node) {
+	n = new(node)
+	n.log = logger.CreateLogger(logger.Default.With(
+		"nodeName", name,
+		"nodeTier", "primary",
+	))
+	n.ws.uri = wsuri
+	if httpuri != nil {
+		n.http = &rawclient{uri: *httpuri}
+	}
+	return
+}
+
+func (n *node) Dial(ctx context.Context) error {
+	if n.dialled {
+		panic("eth.Client.Dial(...) should only be called once during the node's lifetime.")
+	}
+
+	{
+		var httpuri string
+		if n.http != nil {
+			httpuri = n.http.uri.String()
+		}
+		n.log.Debugw("eth.Client#Dial(...)", "wsuri", n.ws.uri.String(), "httpuri", httpuri)
+	}
+
+	{
+		uri := n.ws.uri.String()
+		rpc, err := rpc.DialWebsocket(ctx, uri, "")
+		if err != nil {
+			return err
+		}
+		n.dialled = true
+		n.ws.rpc = rpc
+		n.ws.geth = ethclient.NewClient(rpc)
+	}
+
+	if n.http != nil {
+		uri := n.http.uri.String()
+		rpc, err := rpc.DialHTTP(uri)
+		if err != nil {
+			return err
+		}
+		n.http.rpc = rpc
+		n.http.geth = ethclient.NewClient(rpc)
+	} else {
+		n.log.Warn("ETH_HTTP_URL was not set for the primary node. It is recommended that node operators set ETH_HTTP_URL in addition to ETH_URL to improve performance and reliability")
+	}
+
+	return nil
+}
+
+// RPC wrappers
+
+func (n node) CallContext(ctx context.Context, result interface{}, method string, args ...interface{}) error {
+	n.log.Debugw("eth.Client#Call(...)",
+		"method", method,
+		"args", args,
+		"mode", switching(n),
+	)
+	if n.http != nil {
+		return wrapHTTP(n.http.rpc.CallContext(ctx, result, method, args...))
+	}
+	return wrapWS(n.ws.rpc.CallContext(ctx, result, method, args...))
+}
+
+func (n node) BatchCallContext(ctx context.Context, b []rpc.BatchElem) error {
+	n.log.Debugw("eth.Client#BatchCall(...)",
+		"nBatchElems", len(b),
+		"mode", switching(n),
+	)
+	if n.http != nil {
+		return wrapHTTP(n.http.rpc.BatchCallContext(ctx, b))
+	}
+	return wrapWS(n.ws.rpc.BatchCallContext(ctx, b))
+}
+
+func (n node) EthSubscribe(ctx context.Context, channel interface{}, args ...interface{}) (ethereum.Subscription, error) {
+	n.log.Debugw("eth.Client#EthSubscribe", "mode", "websocket")
+	return n.ws.rpc.EthSubscribe(ctx, channel, args...)
+}
+
+func (n node) Close() {
+	n.ws.rpc.Close()
+}
+
+// GethClient wrappers
+
+func (n node) TransactionReceipt(ctx context.Context, txHash common.Hash) (receipt *types.Receipt, err error) {
+	n.log.Debugw("eth.Client#TransactionReceipt(...)",
+		"txHash", txHash,
+		"mode", switching(n),
+	)
+
+	if n.http != nil {
```

Comment:
I think we could simplify all this if/elses by just saving one `*ethclient.Client`on the node in newNode (with a preference for http if is available) then use `n.geth` directly for all these RPC methods

Reply:
Hmm, I was imagining we might want to have per-call behaviour.

Reply:
yeah I think there could be a default, probably doesn't matter too much fine to leave as is

Reply:
Reality is that we're moving away from websockets completely, let's just minimize friction for ourselves and node ops for now

Per-call++ for now, rethink with Piotr et al shortly